Add extension create command

samcoe opened the pull request:
This PR adds the `extension create` command. It creates the local scaffolding for developing an extension and then outputs instructions on next steps for the user. Does not install the extension locally or publish the extension.

<img width="651" alt="Screen Shot 2021-08-19 at 4 08 39 PM" src="https://user-images.githubusercontent.com/7969779/130155559-8558884b-0417-4700-8227-d9f71c1d358e.png">

Fixes https://github.com/cli/cli/issues/4104
Supersedes https://github.com/cli/cli/pull/4157

samcoe committed:
Add extension create command

samcoe commented on pkg/cmd/extension/command_test.go:
```diff
@@ -161,6 +162,50 @@ func TestNewCmdExtension(t *testing.T) {
+
+				Publish to GitHub with: gh repo create gh-test
+
+				For more information on writing extensions: https://docs.github.com/en/github-cli/github-cli/creating-github-cli-extensions
```
Confirm that this link is correct before merging.

samcoe commented on pkg/cmd/extension/command.go:
```diff
@@ -141,6 +141,39 @@ func NewCmdExtension(f *cmdutil.Factory) *cobra.Command {
+					return err
+				}
+				if io.IsStdoutTTY() {
+					link := "https://docs.github.com/en/github-cli/github-cli/creating-github-cli-extensions"
```
Confirm that this link is correct before merging.

samcoe replied:
Confirmed.

samcoe commented on pkg/cmd/extension/manager.go:
```diff
@@ -249,6 +250,32 @@ func (m *Manager) installDir() string {
+		return err
+	}
+
+	fileTmpl := heredoc.Docf(`
+		#!/bin/bash
+		echo "Hello %[1]s"
+	`, name)
```
Do we want to add anything else to the template?

vilmibm replied:
yes! I think we should have:

- example of checking to see if something is on path (like `ruby`, `python`, or `go`)
- example of passing args through to something (to demonstrate use of `"$@"`)
- example of using `gh api`

samcoe replied:
@vilmibm I added some examples. Let me know what you think.

samcoe committed:
Fix up link

vilmibm reviewed (changes requested):
I was hoping for a survey prompt for the extension name if f they don't provide an argument; I'll also take a friendlier error message when there are no args.

Otherwise, I'm A+ on the code and just want to see some more stuff in the bash script.

samcoe committed:
Better error message

samcoe committed:
Add examples to template

samcoe commented:
@vilmibm I addressed your comments, let me know what you think. I skipped adding a prompt for the extension name since none of the other `extension` commands use prompts and I wanted to keep it consistent.

samcoe committed:
Move link to next line since it is long

vilmibm reviewed (approved):
awesome, thank you! I am happy with the state of things.

mislav commented on pkg/cmd/extension/manager.go:
```diff
@@ -249,6 +250,64 @@ func (m *Manager) installDir() string {
+		# fi
+
+		# Pass arguments through to a new process
+		# exec gh issue list "$@" -R cli/cli
```
I'm not sure if we should show `exec` as an example in the middle of the script. If someone copy-pastes this, nothing after `exec` will run as exec will replace the current process with the subprocess. Since `exec` should always be the last process spawned, but is not strictly needed here, I would just omit it
```suggestion
		# gh issue list "$@" -R cli/cli
```

mislav commented on pkg/cmd/extension/manager.go:
```diff
@@ -249,6 +250,64 @@ func (m *Manager) installDir() string {
+		#   }
+		# '
+		# TEMPLATE='
+		#   {{- range $repo := .data.viewer.repositories.nodes -}}
```
Nice going naming the iteration variable here! 👍

mislav commented on pkg/cmd/extension/manager.go:
```diff
@@ -249,6 +250,64 @@ func (m *Manager) installDir() string {
+		#     {{- printf "name: %[2]s - stargazers: %[3]s\n" $repo.nameWithOwner $repo.stargazerCount -}}
+		#   {{- end -}}
+		# '
+		# exec gh api graphql --template="${TEMPLATE}" --paginate -f query="${QUERY}"
```
Nit: could query go first and template last? So that it follows the order of operations: the requests is made first, then the response is processed using the template
```suggestion
		# exec gh api graphql -f query="${QUERY}"  --paginate --template="${TEMPLATE}"
```

mislav commented on pkg/cmd/extension/manager.go:
```diff
@@ -249,6 +250,64 @@ func (m *Manager) installDir() string {
+		# '
+		# TEMPLATE='
+		#   {{- range $repo := .data.viewer.repositories.nodes -}}
+		#     {{- printf "name: %[2]s - stargazers: %[3]s\n" $repo.nameWithOwner $repo.stargazerCount -}}
```
I think this is a neat example! 🌟

mislav commented on pkg/cmd/extension/manager.go:
```diff
@@ -249,6 +250,64 @@ func (m *Manager) installDir() string {
+		return err
+	}
+
+	initCmd := m.newCommand(exe, "init", "-b", "main", "--quiet", name)
```
Let's not use `-b main` here and respect the person's git settings for default main branch.
```suggestion
	initCmd := m.newCommand(exe, "init", "--quiet", name)
```
We suggest using `-b main` on the docs site because the instructions there are manual; thus, a person following them can change them at invocation time.

mislav commented on pkg/cmd/extension/manager.go:
```diff
@@ -249,6 +250,64 @@ func (m *Manager) installDir() string {
+
+	fileTmpl := heredoc.Docf(`
+		#!/bin/bash
+		set -e
```
Small nit: let's put a space after `set -e` so to separate the typical bash "preamble" from the rest of the script

mislav commented on pkg/cmd/extension/manager.go:
```diff
@@ -249,6 +250,64 @@ func (m *Manager) installDir() string {
+
+		# Determine if an executable is in the PATH
+		# if [ "$(which go)" = "" ]; then
+		#   echo "go not found in PATH"
```
I would use `type` to check whether an executable is in PATH. It doesn't require a subshell and it's a shell builtin as opposed to a separate process. Furthermore, `which` can have subtle differences across platforms.

I would also check for things other than Go because I don't think it's realistic right now that people will be able to make Go extensions without us giving them any resources to do so with. It's much more likely that they will use a scripting language like Ruby or Python.

Finally, I would make sure that all error/warning output goes to stderr, if we're going to set good examples for people's scripts.
```suggestion
		# if ! type -p ruby >/dev/null; then
		#   echo "Ruby not found on the system" >&2
```

mislav commented on pkg/cmd/extension/command.go:
```diff
@@ -141,6 +141,40 @@ func NewCmdExtension(f *cmdutil.Factory) *cobra.Command {
+				if err := m.Create(extName); err != nil {
+					return err
+				}
+				if io.IsStdoutTTY() {
```
Nit: you could invert the condition here and return early to avoid having to indent the rest of the function from here on

mislav reviewed (changes requested):
This is great! Couple of minor but important points

samcoe committed:
Unhide extension commands

samcoe committed:
Address PR  comments

mislav commented on pkg/cmd/extension/manager.go:
```diff
@@ -249,6 +250,64 @@ func (m *Manager) installDir() string {
+		# exec gh api graphql --template="${TEMPLATE}" --paginate -f query="${QUERY}"
+	`, name, "%s", "%v")
+	filePath := filepath.Join(name, name)
+	err = ioutil.WriteFile(filePath, []byte(fileTmpl), 0755)
```
Just had a thought: Windows doesn't have a concept of executable files. When this file gets added to git, will it retain its executable bit? If not, this could be a workaround on Windows that we could execute locally as part of `create`:
```
git add file_name
git update-index --chmod=+x file_name
```

samcoe replied:
I asked @vilmibm to help me check on this and the file created was executable on Windows. I don't know if @vilmibm confirmed that when it was added to git it maintained the executable property though.

mislav replied:
This is unfortunately the trickiest part of publishing an extension on Windows, so if it turns out that it's needed, I would vote to add the `git add && update-index` calls on Windows (or across every platform for simplicity—it would be a no-op on other platforms anyway!)

vilmibm replied:
i just created the extension and ran the `gh-test` file; i didn't try adding to git

samcoe replied:
I went ahead and added `git add ... --chmod=+x` command to the flow.

mislav commented on pkg/cmd/extension/command.go:
```diff
@@ -141,9 +141,43 @@ func NewCmdExtension(f *cmdutil.Factory) *cobra.Command {
 		},
+		&cobra.Command{
+			Use:   "create <name>",
+			Short: "Create a new extension",
```
Just had another thought: should this say `Create a new extension locally` so that it's clear that whatever will be created, it's only on your machine?

It can be either here or in the `Long` description. Or maybe it's not necessary. Just an idea!

mislav reviewed (approved):
Thanks for the quick updates! Looks great 🚀

samcoe committed:
Extension template is executable according to git
